Jenkinsfile

```diff
@@ -1,57 +1,85 @@
-JENKINS_NODE = env.JENKINS_NODE ?: 'master'
+// Node label that specifies on which slave(s) the job should run
+BUILD_SLAVES_LABEL = 'master'
+
+DEFAULT_MAVEN_OPTS = "-Xss256k -Xmx300m -XX:MaxMetaspaceSize=80m -XX:MaxMetaspaceExpansion=10m " +
+                     "-Dmaven.test.failure.ignore=false -XshowSettings:vm " + 
+                     "-XX:+TieredCompilation -XX:TieredStopAtLevel=1"
 
 // Text colours definition
 def GREEN( String msg ) { return "\u001B[32m${msg}\u001B[0m" }
 def YELLOW( String msg ) { return "\u001B[33m${msg}\u001B[0m" }
 def RED( String msg ) { return "\u001B[31m${msg}\u001B[0m" }
 
-def LOGS_PUSH( String str ) {
-  LOGS.push( str )
-}
+def getCause(def build) {
+  while(build.previousBuild) {
+    build = build.previousBuild
+  }
 
-def LOGS_DUMP() {
-  echo LOGS.join("\n")
+  return build.rawBuild.getCause(hudson.model.Cause$UserIdCause)
 }
 
-LOGS = []
-JENKINS_BUILD_USER = "Unknown"
+// About size option, see https://hyper.sh/pricing.html
+def runOnHyper(def name, def size, def mavenOpts, def command) {
+  dir("${WORKSPACE}/") {
+    sh "hyper run --name '${name}' --size=${size} --cidfile ${name}.cid --noauto-volume " +
+      "--restart=no --rm -i -e MAVEN_OPTS='${mavenOpts}' " +
+      "-v \$(pwd):/usr/local/checkstyle/ " +
+      "checkstyle/maven-builder-image:jdk-8u162b12-maven-3.5.3-groovy-2.4.15 " +
+      "bash -c '${command}'"
+  }
+}
 
-// The build pipeline
 pipeline {
 
-  // Specify the Jenkins to to run on
-  agent { node { label "${JENKINS_NODE}" } }
-
-  // Various build and build plugin's options
+  agent {
+    label "${BUILD_SLAVES_LABEL}"
+  }
+    
   options {
     ansiColor('xterm')
   }
 
   stages {
 
-    stage("Build") {
-      agent { node { label "${JENKINS_NODE}" } }
-      steps {
-        echo GREEN("STARTED: Stage 'Noop'")
-        dir("${WORKSPACE}/"){
-          
-          // Running the ls command as a 'no-op' is helpful to make sure that the proper sources were checked out
-          sh "ls -lah"
-          
-          LOGS_PUSH( GREEN("Noop 'ls -lah' task finished") )
+    stage ("Prepare (triggered by hand)") {
+      when { expression { getCause(currentBuild).shortDescription.contains("Started by user") } }
+        steps {
+        echo YELLOW("Triggered by hand, so building for master branch")
```

Comment:
CI build have tendency to fail periodically by reasons unrelated to code changes.
It is always required to have ability to restart PR build by manually.

it should be possible.

Reply:
@romani, sure this is already supported :) 
To re-start the PR build, there are 2 options:
1. Repo team member writes a key phrase 'test this please' in PR's 'conversation' chat
2. PR author can do 'git commit --amend'. Even without any changes, '--amend' updates the commit hash, so the build will be re-triggered regardless the fact that there are no actual code changes

There are also some other options to re-trigger the failed build, but I decided to not implement them for now to not overcomplicate.